src/datasets/arrow_dataset.py

```diff
@@ -3692,21 +3746,28 @@ def add_item(self, item: dict, new_fingerprint: str):
         Returns:
             :class:`Dataset`
         """
-        item_table = InMemoryTable.from_pydict({k: [item[k]] for k in self.features.keys() if k in item})
-        # Cast item
-        schema = pa.schema(self.features.type)
-        item_table = item_table.cast(schema)
-        # Concatenate tables
-        table = concat_tables([self._data, item_table])
+        item_table = InMemoryTable.from_pydict({k: [v] for k, v in item.items()})
+        # We don't call _check_if_features_can_be_aligned here so this cast is "unsafe"
+        dset_features, item_features = _align_features([self.features, Features.from_arrow_schema(item_table.schema)])
+        # Cast and concatenate tables
+        table = concat_tables(
+            [
+                self._data.cast(pa.schema(dset_features.type)) if self.features != dset_features else self._data,
```

Comment:
Why is casting needed here ? I guess this is to re-order the columns if needed ?
Feel free to add a comment in the code since we may forget that this is important x)

Reply:
Nice catch! Yes, I don't think this cast is needed. Let's remove it and see what the CI says :).